src/Kestrel/KestrelConfigBuilder.cs

```diff
@@ -0,0 +1,122 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Server.Kestrel.Core;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal;
+using Microsoft.AspNetCore.Server.Kestrel.Https;
+using Microsoft.AspNetCore.Server.Kestrel.Https.Internal;
+using Microsoft.AspNetCore.Server.Kestrel.Internal;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Server.Kestrel
+{
+    public class KestrelConfigBuilder : IKestrelConfigBuilder
+    {
+        internal KestrelConfigBuilder(KestrelServerOptions options, IConfiguration configuration)
+        {
+            Options = options ?? throw new ArgumentNullException(nameof(options));
+            Configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+
+            Options.ConfigurationBuilder = this;
+        }
+
+        public KestrelServerOptions Options { get; }
+        public IConfiguration Configuration { get; }
+        private IDictionary<string, Action<EndpointConfiguration>> EndpointConfigurations { get; }
+            = new Dictionary<string, Action<EndpointConfiguration>>(0);
+
+        /// <summary>
+        /// Specifies a configuration Action to run when an endpoint with the given name is loaded from configuration.
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="configureOptions"></param>
+        public KestrelConfigBuilder Endpoint(string name, Action<EndpointConfiguration> configureOptions)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            EndpointConfigurations[name] = configureOptions ?? throw new ArgumentNullException(nameof(configureOptions));
+            return this;
+        }
+
+        public void Build()
+        {
+            if (Options.ConfigurationBuilder == null)
+            {
+                // The builder has already been built.
+                return;
+            }
+            Options.ConfigurationBuilder = null;
+
+            var configReader = new ConfigReader(Configuration);
+            
+            foreach (var endpoint in configReader.Endpoints)
+            {
+                var listenOptions = AddressBinder.ParseAddress(endpoint.Url, out var https);
+                listenOptions.KestrelServerOptions = Options;
+                Options.EndpointDefaults(listenOptions);
+
+                HttpsConnectionAdapterOptions httpsOptions = null;
+                if (https)
+                {
+                    httpsOptions = new HttpsConnectionAdapterOptions();
+                    Options.GetHttpsDefaults()(httpsOptions);
+
+                    var certInfo = new CertificateConfig(endpoint.CertConfig);
+                    if (certInfo.IsFileCert)
```

Comment:
Are any errors thrown if a file path and a store location are both specified for the same endpoint?

Reply:
No, I don't want to second guess extraneous data. If we have what we need then there's no reason to fail. Nor do I expect people to mix those two up.

Reply:
I see no reason not to fail if both a file path and a store location is provided.

What's the logic behind choosing file path over the store location when both are specified? If you don't expect people to mix the two, what's wrong with throwing?